src/TenancyServiceProvider.php

```diff
@@ -109,10 +109,12 @@ public function boot(): void
 
         // Queue tenancy
         $this->app['events']->listen(\Illuminate\Queue\Events\JobProcessing::class, function ($event) {
-            if (array_key_exists('tenant_id', $event->job->payload())) {
-                if (! tenancy()->initialized) { // dispatchNow
-                    tenancy()->initialize(tenancy()->find($event->job->payload()['tenant_id']));
-                }
+            $tenantId = $event->job->payload()['tenant_id'] ?? null;
+            if (
+                $tenantId !== null &&
+                (! tenancy()->initialized || tenancy('id') !== $tenantId)
+            ) {
+                tenancy()->initById($tenantId);
```

Comment:
This basically initializes tenancy if the job payload has a tenant_id and the tenant is different from the current tenant (to avoid re-initializing tenancy in case of `dispatchNow`, but to re-initialize tenancy if a new job for a different tenant should be executed), right.

I wonder if this could be refactored somehow. This takes me at least a minute to parse in my head which I don't like. But I can't think of a much better way to write this. Maybe something like this?

```php
$tenantId = $event->job->payload()['tenant_id'] ?? null;

if (! $tenantId) { // The job is not tenant-aware
    return;
}

if (tenancy()->initialized && tenant('id') === $tenantId) { // Tenancy is already initialized for the tenant (e.g. dispatchNow was used)
   return;
}

// Tenancy was either not initialized, or initialized for a different tenant. Therefore, we initialize it for the correct tenant.
tenancy()->initById($tenantId);
```

Reply:
Completely agree @stancl. I've refactored the logic to be much easier to understand what's going on